### PR TITLE
fix: Implements tls-certificates lib v3

### PR DESCRIPTION
**Pull request by gruyaume**
# Description

Here we implement a somewhat minor but API breaking change to the TLS Certificates library with the main changes described below.

### Simplify public interfaces

Here we introduce a V3 of the lib that simplifies the outputs of public interfaces from manipulating complex dicts to simple Python dataclasses objects. This makes the lib much easier to work with and not prone to key type exceptions.

#### Before

```python
def get_outstanding_certificate_requests(self, relation_id: Optional[int] = None) -> List[Dict[str, Union[int, str, List[Dict[str, str]]]]]:
```
Where the dict looked like:

```python
{
                "relation_id": 0,
                "application_name": "tls-certificates-requirer",
                "unit_name": "tls-certificates-requirer/0",
                "unit_csrs": [
                    {
                        "certificate_signing_request": "-----BEGIN CERTIFICATE REQUEST-----...",
                        "is_ca": false
                    }
                ]
            }
```

#### After

```python
def get_outstanding_certificate_requests(self, relation_id: Optional[int] = None) -> List[RequirerCSR]:
```

Where the object attributes can be accessed like

```
requirer_csr.relation_id
requirer_csr.csr
...
```

### Remove unused helpers

We also take this opportunity to get rid of `pfx` packages here as those were only used in Magma which is not distributed anymore.

### Only support for Juju > 3.1
The TLS interface V2 supports both Juju 2.9 and Juju 3.X. The TLS interface V3  here relies on secrets and only works with Juju > 3.1. 

## For reviewers

Most of this lib did not change, I recommend using [this PR](https://github.com/canonical/self-signed-certificates-operator/pull/57/files) to help filter the actual changes we made to the lib.

## Rationale

Some of the TLS provider library's public interfaces used to return complicated dicts that all implementation of provider charms had to understand (ex. self-signed certificates, manual-tls-certificates, etc.). This was quite nasty and caused bugs like [this one](https://github.com/canonical/self-signed-certificates-operator/issues/56) where when trying to get the list of outstanding certificate requests, one of the expected dict key was non existent and caused the charm to crash.

## Checklist

- [ ] My code follows the [style guidelines](/CONTRIBUTING.md) of this project.
- [ ] I have performed a self-review of my own code.
- [ ] I have made corresponding changes to the documentation.
- [ ] I have added tests that validate the behaviour of the software.
- [ ] I validated that new and existing unit tests pass locally with my changes.
- [ ] Any dependent changes have been merged and published in downstream modules.
- [ ] I have bumped the version of any required library.

### CODE REVIEW

File: lib/charms/tls_certificates_interface/v2/tls_certificates.py [hunk 1]
```diff
@@ -308,7 +308,7 @@ def _on_all_certificates_invalidated(self, event: AllCertificatesInvalidatedEven
 
 # Increment this PATCH version before using `charmcraft publish-lib` or reset
 # to 0 if you are raising the major API version
-LIBPATCH = 22
+LIBPATCH = 23
```
Comment (ghislainbourgeois) [inline lib/charms/tls_certificates_interface/v2/tls_certificates.py:311]: This seems like a breaking change unfortunately, should we bump the major version?

File: lib/charms/tls_certificates_interface/v2/tls_certificates.py [hunk 2]
```diff
@@ -423,6 +423,46 @@ def _on_all_certificates_invalidated(self, event: AllCertificatesInvalidatedEven
...
+        self.is_ca = is_ca
+
+    def __eq__(self, other):
+        return (
+            self.relation_id == other.relation_id
+            and self.application_name == other.application_name
+            and self.unit_name == other.unit_name
+            and self.csr == other.csr
+            and self.is_ca == other.is_ca
+        )
```
Comment (ghislainbourgeois) [inline lib/charms/tls_certificates_interface/v2/tls_certificates.py:445]: To avoid breaking the API, we could make these objects be dicts:
```suggestion
class RequirerCSR(collections.UserDict):
    """This class represents a certificate signing request from an interface Requirer."""

    def __init__(
        self, relation_id: int, application_name: str, unit_name: str, csr: str, is_ca: bool
    ):
        self.relation_id = relation_id
        self.application_name = application_name
        self.unit_name = unit_name
        self.csr = csr
        self.is_ca = is_ca
        super().__init__({"relation_id": relation_id, "application_name": application_name, "unit_name": unit_name, "csr": csr, "is_ca": is_ca})
```

I would not use this suggestion as-is but test it out and make a proper design of it. I think an approach like that would prevent us pushing a breaking change, and we could gradually push people to use the new API. A good way of doing that would be to override the dictionnary methods like so:

```
def __getitem__(self, key):
    warnings.warn("Using RequirerCSR as a dict is deprecated", DeprecationWarning, stacklevel=2)
    return super().__getitem__(key)
```

What do you think?

**Review by ghislainbourgeois — CHANGES_REQUESTED** [review]
I have concerns about backwards compatibility and provided some comments.

File: lib/charms/tls_certificates_interface/v2/tls_certificates.py [hunk 3]
```diff
@@ -1446,15 +1464,24 @@ def get_requirer_csrs(
...
-                    }
-                )
-        return unit_csr_mappings
+                for unit_csr in unit_csrs_list:
+                    csr = unit_csr.get("certificate_signing_request")
+                    if not csr:
+                        logger.warning("No CSR found in relation data - Skipping")
+                        continue
+                    ca = unit_csr.get("ca", False)
+                    if not relation.app:
```
Comment (saltiyazan) [inline lib/charms/tls_certificates_interface/v2/tls_certificates.py:1473]: Should we validate this here? this will be validated on every unit, and every csr. I think we should take it out to the first loop and only validate it for every relation.

File: lib/charms/tls_certificates_interface/v2/tls_certificates.py (hunk 2 above)
Comment (gruyaume) [reply lib/charms/tls_certificates_interface/v2/tls_certificates.py:445]: Even if this new object has dict like properties and can be subscribabled(?), it's not a 1to1 with the prior dict so it won't maintain backward compatibility. I'll go ahead and create a new major version of the lib and take this opportunity to also change the requirer side of the interface in the same way

File: lib/charms/tls_certificates_interface/v3/tls_certificates.py [hunk 4]
```diff
@@ -0,0 +1,2002 @@
...
+        self.is_ca = is_ca
+
+    def __eq__(self, other):
+        return (
+            self.relation_id == other.relation_id
+            and self.application_name == other.application_name
+            and self.unit_name == other.unit_name
+            and self.csr == other.csr
+            and self.is_ca == other.is_ca
+        )
```
Comment (Kayra1) [inline lib/charms/tls_certificates_interface/v3/tls_certificates.py:445]:
```suggestion
@dataclass
class RequirerCSR:
    """This class represents a certificate signing request from an interface Requirer."""
        relation_id: int
        application_name: str
        unit_name: str
        csr: str
        is_ca: bool
```
python will give you the same `__eq__` function by default.

File: lib/charms/tls_certificates_interface/v3/tls_certificates.py [hunk 5]
```diff
@@ -0,0 +1,2002 @@
...
+    def __eq__(self, other):
+        return (
+            self.relation_id == other.relation_id
+            and self.application_name == other.application_name
+            and self.csr == other.csr
+            and self.certificate == other.certificate
+            and self.ca == other.ca
+            and self.chain == other.chain
+            and self.revoked == other.revoked
+        )
```
Comment (Kayra1) [inline lib/charms/tls_certificates_interface/v3/tls_certificates.py:478]:
```suggestion
@dataclass
class ProviderCertificate:
    """This class represents a certificate from an interface Provider."""
       
        relation_id: int
        application_name: str
        csr: str
        certificate: str
        ca: str
        chain: List[str]
        revoked: bool
 ```

same as above

**Review by Kayra1 — COMMENTED** [review]
I would highly suggest using dataclasses in this situation. It was designed exactly for this purpose, is native to python since 3.7, and provides a lot of the required boilerplate by default. I've proposed the changes required to use it below.

File: lib/charms/tls_certificates_interface/v3/tls_certificates.py (hunk 4 above)
Comment (gruyaume) [reply lib/charms/tls_certificates_interface/v3/tls_certificates.py]: Done

File: lib/charms/tls_certificates_interface/v3/tls_certificates.py (hunk 5 above)
Comment (gruyaume) [reply lib/charms/tls_certificates_interface/v3/tls_certificates.py]: Done

File: lib/charms/tls_certificates_interface/v3/tls_certificates.py [hunk 6]
```diff
@@ -0,0 +1,2002 @@
...
+juju relate <tls-certificates provider charm> <tls-certificates requirer charm>
+```
+
+"""  # noqa: D405, D410, D411, D214, D416
+
+import copy
+import json
+import logging
+import uuid
+from contextlib import suppress
```
[the remark this answers is not in the document]
Comment (gruyaume) [reply lib/charms/tls_certificates_interface/v3/tls_certificates.py:279]: Done

**Review by ghislainbourgeois — CHANGES_REQUESTED** [review]

File: lib/charms/tls_certificates_interface/v3/tls_certificates.py [hunk 7]
```diff
@@ -0,0 +1,1964 @@
...
+
+@dataclass
+class RequirerCSR:
+    """This class represents a certificate signing request from an interface Requirer."""
+
+    relation_id: int
+    application_name: str
+    unit_name: str
+    csr: str
+    is_ca: bool
```
[the remark this answers is not in the document]
Comment (gruyaume) [reply lib/charms/tls_certificates_interface/v3/tls_certificates.py:438]: Our public interfaces already have default values meaning that creating a new instance of ProviderCertificate for instance will use the default value from the interface. I'm not convinced adding an additional default value would help here.

File: lib/charms/tls_certificates_interface/v3/tls_certificates.py [hunk 8]
```diff
@@ -1767,10 +1762,9 @@ def _on_relation_changed(self, event: RelationChangedEvent) -> None:
         for certificate in provider_certificates:
             if certificate.csr in requirer_csrs:
                 if certificate.revoked:
-                    if JujuVersion.from_environ().has_secrets:
-                        with suppress(SecretNotFoundError):
-                            secret = self.model.get_secret(label=f"{LIBID}-{certificate.csr}")
-                            secret.remove_all_revisions()
+                    with suppress(SecretNotFoundError):
+                        secret = self.model.get_secret(label=f"{LIBID}-{certificate.csr}")
+                        secret.remove_all_revisions()
```
Comment (ghislainbourgeois) [inline lib/charms/tls_certificates_interface/v3/tls_certificates.py]: I am not sure it makes sense to drop Juju 2 support at this time, the current LTS is still 2.9.

**Review by ghislainbourgeois — CHANGES_REQUESTED** [review]

File: lib/charms/tls_certificates_interface/v3/tls_certificates.py [hunk 9]
```diff
@@ -0,0 +1,1964 @@
...
+class ProviderCertificate:
+    """This class represents a certificate from an interface Provider."""
+
+    relation_id: int
+    application_name: str
+    csr: str
+    certificate: str
+    ca: str
+    chain: List[str]
+    revoked: bool
```
[the remark this answers is not in the document]
Comment (gruyaume) [reply lib/charms/tls_certificates_interface/v3/tls_certificates.py:451]: (same comment as above. I'll resolve this one here to focus the discussion in 1 place)

File: lib/charms/tls_certificates_interface/v3/tls_certificates.py (hunk 8 above)
Comment (gruyaume) [reply lib/charms/tls_certificates_interface/v3/tls_certificates.py]: I'm not suggesting to completely drop Juju 2 support. We can still maintain the tls lib v2. But the lib v3 won't support Juju 2.

File: lib/charms/tls_certificates_interface/v3/tls_certificates.py [hunk 10]
```diff
@@ -0,0 +1,1930 @@
...
+        )
+        self.framework.observe(
+            charm.on[relationship_name].relation_broken, self._on_relation_broken
+        )
+        if JujuVersion.from_environ().has_secrets:
+            self.framework.observe(charm.on.secret_expired, self._on_secret_expired)
+        else:
+            self.framework.observe(charm.on.update_status, self._on_update_status)
+
+    def get_requirer_csrs(self) -> List[RequirerCSR]:
```
[the remark this answers is not in the document]
Comment (gruyaume) [reply lib/charms/tls_certificates_interface/v3/tls_certificates.py:1486]: This brings symmetry between the requirer and provider sides of the interface.

For both sides, it can be useful to retrieve the list of requirer CSR's and provider certificates at any event (not just on the custom events).

For this specific use case, let's say you have a charm requiring TLS certs and during a pebble ready event, you want to check whether you already made a request for a CSR, you can call this method and retrieve the list of CSR's.

File: lib/charms/tls_certificates_interface/v3/tls_certificates.py [hunk 11]
```diff
@@ -0,0 +1,1930 @@
...
+                relation_id=relation.id,
+                application_name=self.model.app.name,
+                unit_name=self.model.unit.name,
+                csr=csr,
+                is_ca=ca,
+            )
+            requirer_csrs.append(relation_csr)
+        return requirer_csrs
+
+    def get_provider_certificates(self) -> List[ProviderCertificate]:
```
[the remark this answers is not in the document]
Comment (gruyaume) [reply lib/charms/tls_certificates_interface/v3/tls_certificates.py:1514]: (same as above, let's focus the discussion there)

File: lib/charms/tls_certificates_interface/v3/tls_certificates.py [hunk 12]
```diff
@@ -0,0 +1,1964 @@
...
+        It returns certificates from all relations if relation_id is not specified.
+
+        Returns:
+            List: List of ProviderCertificate objects
+        """
+        provider_certificates = self.get_provider_certificates(relation_id=relation_id)
+        non_revoked_certificates = [
+            certificate for certificate in provider_certificates if not certificate.revoked
+        ]
+        return non_revoked_certificates
```
[the remark this answers is not in the document]
Comment (gruyaume) [reply lib/charms/tls_certificates_interface/v3/tls_certificates.py]: Done

File: lib/charms/tls_certificates_interface/v3/tls_certificates.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (ghislainbourgeois) [reply lib/charms/tls_certificates_interface/v3/tls_certificates.py:438]: It would only be a convenience if we were accepting those objects as public parameters, but we do not seem to, so that is fine with me.

File: lib/charms/tls_certificates_interface/v3/tls_certificates.py [hunk 13]
```diff
@@ -0,0 +1,1964 @@
...
+
+# The unique Charmhub library identifier, never change it
+LIBID = "afd8c2bccf834997afce12c2706d2ede"
+
+# Increment this major API version when introducing breaking changes
+LIBAPI = 3
+
+# Increment this PATCH version before using `charmcraft publish-lib` or reset
+# to 0 if you are raising the major API version
+LIBPATCH = 1
```
[the remark this answers is not in the document]
Comment (gruyaume) [reply lib/charms/tls_certificates_interface/v3/tls_certificates.py]: Fixed

File: lib/charms/tls_certificates_interface/v3/tls_certificates.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (ghislainbourgeois) [reply lib/charms/tls_certificates_interface/v3/tls_certificates.py:1483]: OK, I think it makes sense. Just wanted to be certain before making more stuff public, as it is a pain to release a new major lib version.

File: lib/charms/tls_certificates_interface/v3/tls_certificates.py (hunk 8 above)
Comment (ghislainbourgeois) [reply lib/charms/tls_certificates_interface/v3/tls_certificates.py]: Is it now possible to push updates to older major versions of library? It was problematic in the past, and last I remember it was being worked on.

If we can, then I am fine with this.

File: lib/charms/tls_certificates_interface/v3/tls_certificates.py [hunk 14]
```diff
@@ -0,0 +1,1964 @@
...
+        list_of_csrs = [csr.csr for csr in requirer_csrs]
+        for certificate in provider_certificates:
+            if certificate.csr not in list_of_csrs:
+                self.on.certificate_revocation_request.emit(
+                    certificate=certificate.certificate,
+                    certificate_signing_request=certificate.csr,
+                    ca=certificate.ca,
+                    chain=certificate.chain,
+                )
+                self.remove_certificate(certificate=certificate.certificate)
```
[the remark this answers is not in the document]
Comment (gruyaume) [reply lib/charms/tls_certificates_interface/v3/tls_certificates.py]: I just reverted this to only look at the specific relation. Indeed it didn't make much sense.

File: lib/charms/tls_certificates_interface/v3/tls_certificates.py (hunk 8 above)
Comment (gruyaume) [reply lib/charms/tls_certificates_interface/v3/tls_certificates.py]: Yes, this was fixed

Ref: https://bugs.launchpad.net/snapstore-server/+bug/2024370

Comment (ghislainbourgeois) [reply lib/charms/tls_certificates_interface/v3/tls_certificates.py]: Awesome!

**Review by ghislainbourgeois — APPROVED** [review]

File: .github/workflows/upload.yml [hunk 15]
```diff
@@ -25,3 +25,9 @@ jobs:
           CHARMCRAFT_AUTH: "${{ secrets.CHARMCRAFT_AUTH }}"
         run: |
           charmcraft publish-lib charms.tls_certificates_interface.v2.tls_certificates
+      
+      - name: Publish libs
+        env:
+          CHARMCRAFT_AUTH: "${{ secrets.CHARMCRAFT_AUTH }}"
+        run: |
+          charmcraft publish-lib charms.tls_certificates_interface.v3.tls_certificates
```
Comment (saltiyazan) [inline .github/workflows/upload.yml:33]: Will this actually work? will we have `charmcraft` installed on the system running this?
And should we consider using [the release-libraries](https://github.com/canonical/charming-actions/tree/main/release-libraries) action instead?

Comment (saltiyazan) [reply .github/workflows/upload.yml:33]: I see the upload charm action will install the `charmcraft` snap based on previous runs.

File: .github/workflows/main.yml [hunk 16]
```diff
@@ -74,3 +74,33 @@ jobs:
...
+      - name: Install tox
+        run: pip install tox
+      - name: Setup operator environment
+        uses: charmed-kubernetes/actions-operator@main
+        with:
+          juju-channel: 3/stable
+          provider: microk8s
+          channel: 1.28-strict/stable
+      - name: Run tests using tox
+        run: tox -e integration-v2
```
Comment (saltiyazan) [inline .github/workflows/main.yml:92]:
```suggestion
        run: tox -e integration-v3
```

File: lib/charms/tls_certificates_interface/v3/tls_certificates.py
Comment (saltiyazan) [inline lib/charms/tls_certificates_interface/v3/tls_certificates.py:1]: Why did you remove the `generate_pfx_package` public function? is it not needed to be provided by the V3?

Comment (gruyaume) [reply lib/charms/tls_certificates_interface/v3/tls_certificates.py:1]: As mentioned in this PR description, pfx packages were only used in Magma

File: .github/workflows/upload.yml (hunk 15 above)
Comment (gruyaume) [reply .github/workflows/upload.yml:33]: This was already being done before, he we simply add the v3 lib

Comment (saltiyazan) [reply .github/workflows/upload.yml:33]: Alright, we can do it in a different PR then if needed :)

**Review by saltiyazan — APPROVED** [review]

**Review by Kayra1 — APPROVED** [review]
